ginauth/multitokenmiddleware.go

```diff
@@ -0,0 +1,97 @@
+package ginauth
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/gin-gonic/gin"
+)
+
+// MultiTokenMiddleware Allows for concurrently verifying a token
+// using different middleware implementations. This relies on implementing
+// the GenericAuthMiddleware interface.
+// Only the first detected success will be taken into account.
+// Note that middleware objects don't have to be of Middleware type, that's
+// only one object that implements the interface.
+type MultiTokenMiddleware struct {
+	verifiers []GenericAuthMiddleware
+}
+
+// NewMultiTokenMiddleware builds a MultiTokenMiddleware object from multiple AuthConfigs.
+func NewMultiTokenMiddleware() (*MultiTokenMiddleware, error) {
+	mtm := &MultiTokenMiddleware{}
+	mtm.verifiers = make([]GenericAuthMiddleware, 0)
+
+	return mtm, nil
+}
+
+// Add will append another middleware object (or verifier) to the list
+// which we'll use to check concurrently
+func (mtm *MultiTokenMiddleware) Add(middleware GenericAuthMiddleware) error {
+	if middleware == nil {
+		return fmt.Errorf("%w: %s", ErrInvalidMiddlewareReference, "The middleware reference can't be nil")
+	}
+
+	mtm.verifiers = append(mtm.verifiers, middleware)
+
+	return nil
+}
+
+// AuthRequired is similar to the `AuthRequired` function from the Middleware type
+// in the sense that it'll evaluate the scopes and the token coming from the context.
+// However, this will concurrently evaluate them with the middlewares configured in this
+// struct
+func (mtm *MultiTokenMiddleware) AuthRequired(scopes []string) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		var wg sync.WaitGroup
+
+		res := make(chan error, len(mtm.verifiers))
+
+		wg.Add(len(mtm.verifiers))
+
+		for _, verifier := range mtm.verifiers {
+			go func(v GenericAuthMiddleware, c *gin.Context, r chan<- error) {
+				defer wg.Done()
+
+				cm, err := v.VerifyToken(c, scopes)
+
+				if err != nil {
+					v.SetMetadata(c, cm)
+				}
+
+				r <- err
+			}(verifier, c, res)
+		}
+
+		wg.Wait()
```

Comment:
Do we need to enforce some timeout so if one doesn't return it doesn't hang?

Reply:
Yes, this could probably be handled by each Middleware though.

Reply:
It's now handled by the remote middleware.